### ISSUE TITLE
Why is there seperate train() and val() in Runner

### ISSUE THREAD

**Issue by zyc-ai** [open-mmlab/mmcv#541]
Hi, 

I'm wondering why is there two seperate and very similar function in runner.Runner, instead of a more modularised one like this:

```python
import time
import torch
import mmcv

from mmcv.runner import EpochBasedRunner as _EpochBasedRunner
from mmcv.runner.utils import get_host_info


class EpochBasedRunner(_EpochBasedRunner):
    def _model(self, data_batch, **kwargs):
        if self.batch_processor is None:
            outputs = self.model.train_step(data_batch, self.optimizer, **kwargs)
        else:
            outputs = self.batch_processor(self.model, data_batch, train_mode=self.train_mode, **kwargs)
        if not isinstance(outputs, dict):
            raise TypeError('"batch_processor()" or "model.train_step()"'
                    ' must return a dict')
        return outputs

    def _run(self, data_loader, mode, **kwargs):
        self.mode = mode
        self.model.train() if self.train_mode else self.model.eval()
        self.data_loader = data_loader
        self._max_iters = self._max_epochs * len(self.data_loader)
        self.call_hook(f'before_{mode}_epoch')
        time.sleep(2)  # Prevent possible deadlock during epoch transition
        for i, data_batch in enumerate(self.data_loader):
            self._inner_iter = i
            self.call_hook(f'before_{mode}_iter')
            outputs = None
            if self.train_mode:
                outputs = self._model(data_batch, **kwargs)
            else:
                with torch.no_grad():
                    outputs = self._model(data_batch, **kwargs)
            if outputs is None:
                raise ValueError(f'Unable to get any output, it is possible the mode {self.mode} is illegal')
            if 'log_vars' in outputs:
                self.log_buffer.update(outputs['log_vars'],
                                       outputs['num_samples'])
            self.outputs = outputs
            self.call_hook(f'after_{mode}_iter')
            self._iter += 1

        self.call_hook(f'after_{mode}_epoch')
        if self.train_mode:
            self._epoch += 1

    def run(self, data_loaders, workflow, max_epochs, **kwargs):
        """Start running.

        Args:
            data_loaders (list[:obj:`DataLoader`]): Dataloaders for training
                and validation.
            workflow (list[tuple]): A list of (phase, epochs) to specify the
                running order and epochs. E.g, [('train', 2), ('val', 1)] means
                running 2 epochs for training and 1 epoch for validation,
                iteratively.
            max_epochs (int): Total training epochs.
        """
        assert isinstance(data_loaders, list)
        assert mmcv.is_list_of(workflow, tuple)
        assert len(data_loaders) == len(workflow)

        self._max_epochs = max_epochs
        for i, flow in enumerate(workflow):
            mode, epochs = flow
            if mode == 'train':
                self._max_iters = self._max_epochs * len(data_loaders[i])
                break

        work_dir = self.work_dir if self.work_dir is not None else 'NONE'
        self.logger.info('Start running, host: %s, work_dir: %s',
                         get_host_info(), work_dir)
        self.logger.info('workflow: %s, max: %d epochs', workflow, max_epochs)
        self.call_hook('before_run')

        while self.epoch < max_epochs:
            for i, flow in enumerate(workflow):
                mode, epochs = flow
                if not isinstance(mode, str):  # self.train()
                    raise TypeError(
                        'mode in workflow must be a str, but got {}'.format(
                            type(mode)))

                for _ in range(epochs):
                    self.train_mode = True if 'train' in mode else False
                    if mode == 'train' and self.epoch >= max_epochs:
                        break
                    self._run(data_loaders[i], 'mode', **kwargs)

        time.sleep(1)  # wait for some hooks like loggers to finish
        self.call_hook('after_run')

```

**Comment by hellock** [comment]
The two methods can definitely be unified with a single one. For simplicity and extensibility we just leave them there, but I think using a method to wrap `self.model.train_step()` and `self.batch_processor()` is clear and simple.

**Comment by zyc-ai** [comment]
Thanks for the reply. The reason why I filed this issue is that when I tried to add another mode, I must write my own runner which seems not quite right. Besides this, I also have to copy the entire train/eval function and edit the `call_hook` to match the mode. And with the above modification, the only thing I have to do is to add mode in my config file, and add some hooks so that it could have different training behavior on-the-fly.
